luigi/contrib/external_program.py

```diff
@@ -131,6 +148,43 @@ def run(self):
                 tmp_stderr.close()
                 tmp_stdout.close()
 
+    @contextmanager
+    def proc_with_tracking_url_context(self, proc_args, proc_kwargs):
+        time_to_sleep = 0.5
+
+        file_to_write = proc_kwargs.get('stdout')
+        proc_kwargs.update({'stdout': subprocess.PIPE})
+        main_proc = subprocess.Popen(proc_args, **proc_kwargs)
+        pipe_to_read = main_proc.stdout
+
+        def _track_url_by_pattern():
+            """
+            Scans the pipe looking for a passed pattern, if the pattern is found, `set_tracking_url` callback is sent.
+            If tmp_stdout is passed, also appends lines to this file.
+            """
+            pattern = re.compile(self.tracking_url_pattern)
+            for new_line in iter(pipe_to_read.readline, ''):
+                if new_line:
+                    if file_to_write:
+                        file_to_write.write(new_line)
+                    match = re.search(pattern, new_line.decode('utf-8'))
+                    if match:
+                        self.set_tracking_url(match.group(1))
+                else:
+                    sleep(time_to_sleep)
+
+        track_proc = Process(target=_track_url_by_pattern)
+        try:
+            track_proc.start()
+            with ExternalProgramRunContext(main_proc):
+                yield main_proc
+        finally:
+            # need to wait a bit to let the subprocess read the last lines
+            track_proc.join(time_to_sleep * 2)
+            if track_proc.is_alive():
+                track_proc.terminate()
+            pipe_to_read.close()
```

Comment:
Thanks for updating this part of the code. Since it's always easy to make a mistake, would you mind trying to run this for a week in production for you?

Reply:
Yup, did it with the latest changes, I'll let you know how it goes